sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExec.scala

```diff
@@ -0,0 +1,570 @@
+  }
+
+  override def hasNext(): Boolean = groupBatches.nonEmpty || wrapped.hasNext
+
+  override def next(): ColumnarBatch = {
+    if (groupBatches.nonEmpty) {
+      groupBatches.dequeue().getColumnarBatch()
+    } else {
+      val batch = wrapped.next()
+      if (batch.numRows() > 0 && batch.numCols() > 0 && partitionSpec.nonEmpty) {
+        GpuSemaphore.acquireIfNecessary(TaskContext.get())
+        val partitionIndices = partitionSpec.indices
+        // 1) Calculate the split indices via cudf.Table.groupBy and aggregation Count.
+        //   a) Compute the count number for each group in a batch, including null values.
+        //   b) Restore the original order (Ascending with NullsFirst) defined in the plan,
+        //      since cudf.Table.groupBy will break the original order.
+        //   c) The 'count' column can be used to calculate the indices for split.
+        val cntTable = withResource(GpuProjectExec.project(batch, partitionSpec)) { projected =>
+          withResource(GpuColumnVector.from(projected)) { table =>
+            table
+              .groupBy(partitionIndices:_*)
+              .aggregate(Aggregation.count(true).onColumn(0))
+          }
+        }
+        val orderedTable = withResource(cntTable) { table =>
+          table.orderBy(partitionIndices.map(id => Table.asc(id, true)): _*)
+        }
+        val (countHostCol, numRows) = withResource(orderedTable) { table =>
+          // Yes copying the data to host, it would be OK since just copying the aggregated
+          // count column.
+          (table.getColumn(table.getNumberOfColumns - 1).copyToHost(), table.getRowCount)
+        }
+        val splitIndices = withResource(countHostCol) { cntCol =>
+          // Verified the type of Count column is integer, so use getInt
+          // Also drop the last count value due to the split API definition.
+          (0L until (numRows - 1)).map(id => cntCol.getInt(id))
+        }.scan(0)(_+_).tail
+
+        // 2) Split the table when needed
+        val resultBatch = if (splitIndices.nonEmpty) {
+          // More than one group, split it and close this incoming batch
+          withResource(batch) { cb =>
+            withResource(GpuColumnVector.from(cb)) { table =>
+              withResource(table.contiguousSplit(splitIndices:_*)) { tables =>
+                // Return the first one and enqueue others
+                val splitBatches = tables.safeMap(table =>
+                  GpuColumnVectorFromBuffer.from(table, GpuColumnVector.extractTypes(batch)))
+                groupBatches.enqueue(splitBatches.tail.map(sb =>
+                  SpillableColumnarBatch(sb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)): _*)
+                splitBatches.head
+              }
+            }
+          }
+        } else {
+          // Only one group, return it directly
+          batch
+        }
+        // Shall we release the semaphore here ?
+        GpuSemaphore.releaseIfNecessary(TaskContext.get())
```

Comment:
We only want to release it if the data is off of the GPU.  This is why we have a callback in the cudf API that writes the data out over the wire so you know when it is done copying the data to the CPU before it starts to write it out over the wire.

Reply:
Got it, thanks, i think here we should release the semaphore, since the data has been off of the GPU already.

Reply:
Sorry no.  The batch being returned (resultBatch) is still on the GPU so you cannot release the semaphore until it is off the GPU. You also do not want to hold the lock while doing any I/O which is why I said that we added in a callback in the cudf API.

https://github.com/NVIDIA/spark-rapids/blob/d865883ee35034f119e4e014ad1e89d354688907/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala#L426

Is the place where the semaphore is released properly and this should be removed.

Reply:
Shall i also remove this line for semaphore acquiring? https://github.com/NVIDIA/spark-rapids/blob/d865883ee35034f119e4e014ad1e89d354688907/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExec.scala#L117 

I release this semaphore here because i acquire one before executing the computation as above.

Reply:
You are right you don't need to acquire the semaphore either.  The semaphore is about GPU memory, not computation.  You grab it when you are going to put something on the GPU and you release it when you are off the GPU or everything on the GPU can be spilled.  Typically you don't need to worry about this because there are only a few places where this happens, but this is one of those cases.

Reply:
ok, removed, thanks.